sysconfig/cloudinit.go

```diff
@@ -88,3 +91,87 @@ func configureCloudInit(opts *Options) (err error) {
 	}
 	return err
 }
+
+// CloudInitState represents the various cloud-init states
+type CloudInitState int
+
+var (
+	// the (?m) is needed since cloud-init output will have newlines
+	cloudInitStatusRe = regexp.MustCompile(`(?m)^status: (.*)$`)
+
+	cloudInitSnapdRestrictFile = "/etc/cloud/cloud.cfg.d/zzzz_snapd.cfg"
+)
+
+const (
+	// CloudInitDisabledPermanently is when cloud-init is disabled as per the
+	// cloud-init.disabled file.
+	CloudInitDisabledPermanently CloudInitState = iota
+	// CloudInitRestrictedBySnapd is when cloud-init has been restricted by
+	// snapd with a specific config file.
+	CloudInitRestrictedBySnapd
+	// CloudInitUntriggered is when cloud-init is disabled because nothing has
+	// triggered it to run, but it could still be run.
+	CloudInitUntriggered
+	// CloudInitDone is when cloud-init has been run on this boot.
+	CloudInitDone
+	// CloudInitEnabled is when cloud-init is active, but not necessarily
+	// finished. This matches the "running" and "not run" states from cloud-init
+	// as well as any other state that does not match any of the other defined
+	// states, as we are conservative in assuming that cloud-init is doing
+	// something.
+	CloudInitEnabled
+	// CloudInitErrored is when cloud-init tried to run, but failed or had invalid
+	// configuration.
+	CloudInitErrored
+)
+
+// CloudInitStatus returns the current status of cloud-init. Note that it will
+// first check for static file-based statuses first through the snapd
+// restriction file and the disabled file before consulting
+// cloud-init directly through the status command.
+// Also note that in unknown situations we are conservative in assuming that
+// cloud-init may be doing something and will return CloudInitEnabled when we
+// do not recognize the state returned by the cloud-init status command.
+func CloudInitStatus() (CloudInitState, error) {
+	// if cloud-init has been restricted by snapd, check that first
+	snapdRestrictingFile := filepath.Join(dirs.GlobalRootDir, cloudInitSnapdRestrictFile)
+	if osutil.FileExists(snapdRestrictingFile) {
+		return CloudInitRestrictedBySnapd, nil
+	}
+
+	// if it was explicitly disabled via the cloud-init disable file, then
+	// return special status for that
+	disabledFile := filepath.Join(dirs.GlobalRootDir, "etc/cloud/cloud-init.disabled")
```

Comment:
Nitpick, there's a constant for the other file above, we could have a constant for this file as well. The other constant starts with `/etc` while this one starts with `etc`. I would make that consistent.